Polymorphic 'this' type

ahejlsberg opened the pull request:
This PR implements polymorphic typing of `this` for classes and interfaces as inspired by #3694. The new features are:

* The type of `this` in an expression within a non-static class or interface member is considered to be an instance of *some class that derives from the containing class* as opposed to simply an instance of the containing class.
* The `this` keyword can be used in a type position within a non-static class or interface member to reference the type of `this`.
* When a class or interface is referenced as a type, all occurrences of the `this` type within the class (including those inherited from base classes) are replaced with the type itself.

This feature makes patterns such as [fluent interfaces](https://en.wikipedia.org/wiki/Fluent_interface) and [covariant return types](https://en.wikipedia.org/wiki/Covariant_return_type) much easier to express and implement. Languages such as C++, Java, and C# use a somewhat involved generic pattern to emulate this feature, [as described here](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern).

An example:

```typescript
class A {
    foo() {
        return this;
    }
}

class B extends A {
    bar() {
        return this;
    }
}

var b: B;
var x = b.foo().bar();  // Fluent pattern works, type of x is B
```

In a non-static member of a class or interface, `this` in a type position refers to the type of `this`. For example:

```typescript
class Entity {
    clone(): this {
        // Code to clone Entity
    }
    equals(other: this): boolean {
        // Code to compare Entity with another Entity
    }
}
```

Each subclass of the above `Entity` will have a `clone` method that returns an instance of the subclass, and an `equals` method that takes another instance of the subclass.

The `this` type is a subtype of and assignable to the instance type of the containing class or interface, *but not vice-versa* (because `this` might actually be a subclass). Thus, certain code patterns that previously compiled may now need an extra type annotation:

```typescript
class A {
    getInstance() {  // Should be getInstance(): A
        return this;
    }
}

class B extends A {
    getInstance() {
        return new B();
    }
}
```

The example above now errors because the inferred return type of `getInstance` in `A` is `this` and the inferred type of `getInstance` in `B` is `B`, which is not assignable to `this`. The fix is to add a return type annotation for `getInstance` in `A`.

The polymorphic `this` type is implemented by providing every class and interface with an implied type parameter that is constrained to the containing type itself (except when the compiler can can tell that `this` is never referenced within the class or interface). That in particular turns a lot of previously non-generic classes into generic equivalents, causing more symbols and types to be created due to generic instantiation. The observed cost in batch compile time ranges from 0% in code that uses no classes to 6-7% in very class-heavy code.

Note that this PR specifically doesn't aim to implement other parts of #3694 such as `this` type annotations for functions. Those will be covered by other PRs if we choose to implement them.

jbondc commented:
:star: Some quick thoughts:
- It is possible to show more information about the ```this``` type?
e.g.
```
class A {
    foo() {
        return this; // this <A>
    }
}
class B extends A {
    bar() {
        return this; // this <B & A>
    }
}
```
Would seem helpful.

- Really interesting about ```isIndependentType()```, I was similarly doing something similar for #3496 but called it a 'terminating type' / isTerminatingType().

- The 'this type' could be called a 'self reference' type and have a special 'this' flag on it.
Think it means could use:
```
    else if (type.flags & TypeFlags.SelfReferenceType) {  
            if (type.flags & TypeFlags.ThisType) {  
                  writer.writeKeyword("this ");  
             }
             writer.writeKeyword("<"+ type.referenceName + ">");  
     }
```
Or something like that.

ahejlsberg commented:
@jbondc Agreed, it might be good to show which class a particular `this` type belongs to in error messages and information produced by the language service.

errorx666 commented:
:clap:

RyanCavanaugh commented on src/compiler/checker.ts:
```diff
@@ -2882,6 +2887,28 @@ namespace ts {
+                        for (let node of baseTypeNodes) {
+                            if (isSupportedExpressionWithTypeArguments(node)) {
+                                let baseSymbol = resolveEntityName(node.expression, SymbolFlags.Type, /*ignoreErrors*/ true);
+                                if (!baseSymbol || !(baseSymbol.flags & SymbolFlags.Interface) || getDeclaredTypeOfClassOrInterface(baseSymbol).thisType) {
```
How does this handle the case where an interface `extends` a `class` that uses `this` ?

ahejlsberg replied:
If one of the base types is a class (i.e. not an interface) we simply return true (because every class has a `this` type). We only return false from this function if the interface itself doesn't reference this, if every base type is an interface, and if no base interface has a `this` type.

RyanCavanaugh commented on src/compiler/checker.ts:
```diff
@@ -4843,9 +4981,26 @@ namespace ts {
-            function typesRelatedTo(sources: Type[], targets: Type[], reportErrors: boolean): Ternary {
+            //function typesRelatedTo(sources: Type[], targets: Type[], reportErrors: boolean): Ternary {
+            //    let result = Ternary.True;
+            //    for (let i = 0, len = sources.length; i < len; i++) {
```
Remove commented block?

ahejlsberg replied:
Oops, yes.

RyanCavanaugh commented on src/compiler/types.ts:
```diff
@@ -376,6 +376,7 @@ namespace ts {
         OctalLiteral =      0x00010000,  // Octal numeric literal
         Namespace =         0x00020000,  // Namespace declaration
         ExportContext =     0x00040000,  // Export context (initialized by binding)
+        ContainsThis =      0x00080000,  // Contains reference to "this"
```
Clarify that this flag only applies to interfaces

RyanCavanaugh commented on tests/baselines/reference/thisInInvalidContexts.errors.txt:
```diff
@@ -49,6 +50,8 @@ tests/cases/conformance/expressions/thisKeyword/thisInInvalidContexts.ts(45,9):
     function genericFunc<T>(x: T) { }
     genericFunc<this>(undefined);  // Should be an error
+                ~~~~
+!!! error TS2526: 'this' type is available only in a non-static member of a class or interface.
```
:+1:

RyanCavanaugh commented:
Need to un-comment the fourslash tests that were edited

ahejlsberg commented:
@RyanCavanaugh Yes, @mhegazy is looking at fixes for the failing fourslash tests.

jbondc commented on tests/baselines/reference/typeOfThis.errors.txt:
```diff
@@ -119,6 +145,8 @@ tests/cases/conformance/expressions/thisKeyword/typeOfThis.ts(130,16): error TS1
             var p: MyGenericTestClass<T, U>;
+                ~
+!!! error TS2403: Subsequent variable declarations must have the same type.  Variable 'p' must be of type 'this', but here has type 'MyGenericTestClass<T, U>'.
             p = v;
```
Ideally these should not fail right? But only on instantiation.

ahejlsberg replied:
No, they should fail. The first declaration infers type `this` for `v`, the second declaration specifies type `MyGenericTestClass<T, U>` for `v`. Those are not identical types, so error.

DanielRosenwasser commented:
@ahejlsberg I spoke with @ryancavanaugh, @weswigham, and @danquirk about this, and I think there are some issues with the meaning of `this` in a type literal. Take the following:

```TypeScript
interface Thing {
    getResource(): {
        methodA(): this;
        methodB(): this;
    }
}
```

In the current implementation, `this` refers to the type assignable to `Thing`; *however*, this doesn't accurately reflect the type of `this` when `methodA` is invoked off of its returned object. This could be *very* confusing for users because the intuition is that the method literally returns the expression `this`.

For either one semantics of the `this` type, a workaround exists to get the alternative semantics by extracting the resource type into its own interface.

However, if `this` refers to the enclosing type literal, then an *easier* fix exists by adding a type parameter to the method:

```TypeScript
    interface Thing {
    getResource<OwnerT extends Thing>(): {
        a(): OwnerT;
        b(): this;
    }
}
```

I think that we should go with the semantics that most-closely matches the runtime: in this case, the `this` type should refer to the type returned by `getResource()`.

ahejlsberg commented:
@DanielRosenwasser You're incorrect about the current behavior, your example is actually an error:

```TypeScript
interface Thing {
    getResource(): {
        methodA(): this;  // Error
        methodB(): this;  // Error
    }
}
```

The error reported is `'this' type is available only in a non-static member of a class or interface` and I think that is the behavior we want. If at some point we decide to make it possible to reference `this` of a function or method (perhaps because the function declared a `this` type parameter as speculated about in #3694) the above would become valid, but for now it should be an error.

In order to reference an outer `this` in an object type literal you have to introduce a generic type and capture this as an argument:

```TypeScript
interface Resource<T> {
    methodA(): T;
    methodB(): T;
}

interface Thing {
    getResource(): Resource<this>;
}
```

Note, however, that you *can* capture the outer this in an object literal:

```typescript
class A {
    foo() {
        return { x: this, f: () => this };
    }
}
```

In order to write down the type inferred for the above, you'd have to resort to a temporary type:

```typescript
interface Foo<T> { x: T; f(): T; }

class A {
    foo(): Foo<this> {
        return { x: this, f: () => this };
    }
}
```

jbondc commented:
So close, @ahejlsberg any thoughts here? 
```
enum D {
	a
}
class B { b() { } }
class C { c() { } }

class Test<T> {
	constructor() {
		// do mixing of things here
	}
	me(): T & this { return this }
}

type mix = D & B & C;

let mixin = new Test<mix>();
```

jbondc commented:
This could be quite nice:
```
class Test<T> {
	constructor(...T) {
		// do mixing of things here
	}
	me(): ...T & this { return this } // on instantiation becomes Test<D & B & C> & this
}

let mixin = new Test(D, B, C);
```
Though not sure of the plumbing needed for it.

ahejlsberg commented:
@jbondc I think the `this` type and mixins are largely orthogonal issues. I'd take the discussion up on one of the many mixin suggestion threads we already have going on and keep this one focused to the semantics and implementation of the `this` type.

DanielRosenwasser commented:
> your example is actually an error

I actually was asking others about the implementation and it was our understanding that it was allowed; sorry about that. I didn't get the chance to try the branch out yesterday.

> Note, however, that you *can* capture the outer this in an object literal:
> ```TypeScript
class A {
    foo() {
        return { x: this, f: () => this };
    }
}
```

But the .d.ts emit for that is currently not valid:

```TypeScript
declare class A {
    foo(): {
        x: this;
        f: () => this;
    };
}
```

If I try to reuse that `.d.ts` file, I get:

```
foo.d.ts(3,12): error TS2526: 'this' type is available only in a non-static member of a class or interface.
foo.d.ts(4,18): error TS2526: 'this' type is available only in a non-static member of a class or interface.
```

ahejlsberg commented:
@DanielRosenwasser @mhegazy The .d.ts issue is tricky. Personally I think this is one of those situations where we should issue an error if a .d.ts is requested with guidance that a type annotation is required.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -2982,6 +3020,81 @@ namespace ts {
+            return true;
+        }
+
+        // A type is considered independent if it is a built-in type keyword, an array with an element type that is
```
Maybe clarify that `this` is not a built-in type keyword.

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -3026,44 +3141,57 @@ namespace ts {
+            let source = resolveDeclaredMembers(type.target);
+            let typeParameters = concatenate(source.typeParameters, [source.thisType]);
+            let typeArguments = type.typeArguments && type.typeArguments.length === typeParameters.length ?
+                type.typeArguments : concatenate(type.typeArguments, [type]);
```
Is `type` the `this` type?

DanielRosenwasser replied:
Actually, would it be possible to just make it independent of the type argument list?

ahejlsberg replied:
I assume you mean is `type` the type argument for the `this` type parameter. If so, yes. We don't want to keep it separate from the type argument list because we cache instantiations based on their type arguments.

yuit replied:
Have the similar question, what is the `type`?

ahejlsberg replied:
`type` is the type reference this function is resolving. Unless a type reference explicitly includes a type argument for `this` (which will appear as an extra type argument beyond what the type parameter list calls for), the type reference *itself* is added as a type argument for this. [See comment here for an explanation](https://github.com/Microsoft/TypeScript/pull/4910/files#diff-4b8bd1eea29904f1be39cd864e1a45c0R1862).

yuit commented on src/compiler/checker.ts:
```diff
@@ -2882,6 +2887,31 @@ namespace ts {
             }
         }
 
+        // Returns true if the interface given by the symbol is free of "this" references. Specifically, the result is
```
nit: use jsdoc. I know we are not really consistent around comment format but it is very helpful to get jsdoc in the language service

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -3760,22 +3888,24 @@ namespace ts {
-                    for (let i = 0; i < types.length; i++) {
-                        if (i > 0) {
-                            result += ",";
+            if (types) {
```
Why did this guard become necessary?

ahejlsberg replied:
Because a type argument list may now be empty (i.e. undefined) in a type reference. There's a comment on the `TypeReference` interface in types.ts that explains how it works.

DanielRosenwasser commented on src/compiler/diagnosticMessages.json:
```diff
@@ -1652,6 +1652,10 @@
         "category": "Error",
         "code": 2525
     },
+    "'this' type is available only in a non-static member of a class or interface.": {
```
`A 'this' type annotation is only allowed in a non-static member of a class or interface.`

ahejlsberg replied:
No, this isn't about type annotations, this is about references to the `this` type (which might or might not occur in type annotations, for example the reference could be in a type assertion).

DanielRosenwasser replied:
Then perhaps just make it `A 'this' type is only allowed...` because that follows the intended naming conventions.

Artazor commented:
As I understand, polymorphic `this` type will greatly simplify the following code:
```TypeScript
// interface Cloneable<T extends Cloneable<T>> {
interface Cloneable<T extends Cloneable<any>> {
      clone(): T;
}
```
turns it into
```TypeScript
interface Cloneable {
       clone(): this;
}
```
It's cool!

However I looks like that it solves only a half of a problem (ok, let it be 90% in common practices)

Am I wight that the system
```TypeScript
// interface Vertex<V extends Vertex<V,E>, E extends Edge<V,E>> {
interface Vertex<V extends Vertex<any,any>, E extends Edge<any,any>> {
      incoming: E[];
      outgoing: E[];
}

//interface Edge<V extends Vertex<V,E>, E extends Edge<V,E>> {
interface Edge<V extends Vertex<any,any>, E extends Edge<any,any>> {
     from: V;
     to: V;
}

class City extends Vertex<City, Road> { 
     constructor(public name: string){
          this.incoming = [];
          this.outgoing = [];
     }
}
class Road extends Edge<City, Road> { 
    constructor(public distance: number, public from: City, public to: City) {
           from.outgoing.push(this);
           to.incoming.push(this);
    }
}

var LA = new City('Los Angeles');
var SF = new City('San Francisco');
var hyperloop = [new Road(558.68, LA, SF), new Road(558.68, SF, LA)]

console.log(hyperloop[0].from.outgoing[0].to.name) // stronly typed

```

still could not benefit from `this` type without introducing _high order types_ into TypeScript ?

I could imagine the following implementation
```TypeScript
// interface Vertex<E<T> extends Edge<T>>  -or-
interface Vertex<E extends Edge> {
      incomming: E<this>[];
      outgoing: E<this>;
}

// interface Edge<V<T> extends Vertex<T>> -or-
interface Edge<V extends Vertex> {
     from: V<this>;
     to: V<this>;
}

class City extends Vertex<Road> { ... }
class Road extends Edge<City> { ... }

```
where for high order types the constraint 
```<G1 extends G2>```
could be formulated as 

`G1<T> extends G2<T>` recursively for every T


Am I right? Or there is a simpler solution that I've missed?

Artazor commented:
I mean that `this` type alone solves only direct type recursion problem. Mutually recursive constrains are still not expressible.
@ahejlsberg ?

ahejlsberg commented:
@Artazor Correct, the polymorphic `this` helps in the simple (and common) case, but the general solution likely requires higher-order types of some form since the dependent types can't otherwise refer to each other's `this`.

GoToLoop commented:
Would polymorphic `this` fix the inheritance case below btW?

```ts
class Foo {
    foo() { return this; }
    //foo(): this { return this; } //
}

class FooBar extends Foo {
}

let q = new FooBar;  // q is of type FooBar.
let w = q.foo();  // but w is of type Foo instead of FooBar.
```

ahejlsberg commented:
@GoToLoop Yes, in your example the type of `w` will be `FooBar`.

GoToLoop commented:
Very nice! Always wanted inherited methods returning `this` as of current class type instead of re-writing the whole method!
I hope this new feature is coming for current 1.6 rather than 1.7. I'll be awaiting for it!

ahejlsberg commented:
Thanks. We already shipped 1.6, but it should be in 1.7. Meanwhile, if you can't wait, once this PR gets merged you can always install a nightly build.

ahejlsberg commented:
Latest commits implement these fixes:

* Ensure that `this` can be referenced in call, construct, and index signature members.
* In declaration files, emit references to an inaccessible `this` as references to the containing class.

Consider the example:

```typescript
class C {
    foo() {
        return { x: this };
    }
}
```

The type inferred for `foo` contains an inaccessible reference to `this`, so when a declaration file is generated, the reference is instead emitted as a reference to the containing class:

```typescript
declare class C {
    foo(): {
        x: C;
    }
}
```

We could alternatively issue an error, but I think this solution is pretty reasonable.

Artazor commented:
@ahejlsberg, just as a fantasy how mutual polymorphic types could be expressed (without high order types)

Imagine a new hypothetical construct — the _coherent-module_, that looks like a module with formal type parameters and can be used for joining the coupled polymorphic types under the same visibility scope (cohernet modules are completely different from the generic packages in Ada):

```TypeScript
module graph<Vertex, Edge> {  // Vertex and Edge - are open polymorphic coherent types

  export class Vertex {
    incomming: Edge[];
    outgoing: Edge[];
  }

  export class Edge {
    from: Vertex;
    to: Vertex;
  }

}
```
We can ignore it's formal parameters and use it like a normal module. Thus `graph.Edge` and `graph.Vertex` are normal (terminal) types without any open polymorphic cross-references.

However, if we will use the following form:

```TypeScript
module geo<City, Road> extends graph<City, Road> {
  class City {   // no need to write that City extends graph.Vertex
      name: string;
  }
  class Road {
      distance: number;
  }
}
```

we will find that the classes geo.Road and geo.City have their actual structures like if they were declared in the following way:
```TypeScript
module geo {
  class City {
      incoming: Road[];
      outgoing: Road[];
      name: string;
  }
  class Road {
      from: City;
      to: City;  
      distance: number;
  }
}
```

Thus, the coherent module inheritance will keep coherence between polymorphic cross-references.


Also it looks like that
```TypeScript 
interface Cloneable {
      clone(): this;
}
class Model implements Cloneable { ... }
```

can be expressed with cohernet modules semantics as

```TypeScript
module cloneable<Cloneable> {
     export interface Cloneable {
          clone(): Cloneable;
     }
}

module model<Model> extends cloneable<Model> {
    export class Model {
          ...
    }
}
```

Is it worth to convert this thoughts into proposal, or there is no real chances to get anything close to be implemented in TS?

ahejlsberg commented:
@Artazor After your first comments I did have similar thoughts about some form of "container" that aggregates the dependent types. It is definitely an interesting concept. That said, it feels like it carries too high of a conceptual load for what it brings. I think it would be very hard to explain when you need it and how to use it properly.

ahejlsberg commented:
@mhegazy @DanielRosenwasser @RyanCavanaugh My latest commits have fixed quick info, added tests, and addressed CR feedback. All that remains is the commented out fourslash tests, but that is a minor issue we can cover in a separate PR. I think this one is ready to merge. Look it over and let me know.

Artazor commented:
@ahejlsberg, I'd say that the conceptual overload here is not far more than for already familiar idiom
```TypeScript
class Base<T extends Base<T>> { ... }  
class More<T extends More<T>> extends Base<T> { ... }  // open inheritance
class Goal extends More<Goal> { ... }  // terminal inheritace
```
adopted by Java and C#

That said, I agree that the generalisation of this pattern for coherent covariance for inherited classes (I'd propose the term _coheritance_) is not very popular. And here are my points:
* I believe that it's popularity is low _due to the monstrously wordy syntax_ and not due to the high conceptual load. 
* **When you need it:** instead of delivering individual classes within your library you want to provide the entire _design pattern_ where each class has its well-defined role:
    * these classes can't be consumed separately, the user needs to complete them all to benefit from their relations; 
    * at the same time you want to take advantage from the static typing.



Let's consider three dependent classes: `A`, `B` and `C` that we want to be coherently inheritable. In Java, their signatures would look like

```TypeScript
class A<TA extends A<TA,TB,TC>, TB extends B<TA,TB,TC>, TC extends C<TA,TB,TC>> { ... } 
class B<TA extends A<TA,TB,TC>, TB extends B<TA,TB,TC>, TC extends C<TA,TB,TC>> { ... } 
class C<TA extends A<TA,TB,TC>, TB extends B<TA,TB,TC>, TC extends C<TA,TB,TC>> { ... }

//open coheritance
class MoreA<
    TA extends MoreA<TA,TB,TC>, 
    TB extends MoreB<TA,TB,TC>, 
    TC extends MoreC<TA,TB,TC>
> extends A<TA, TB, TC> { ... } 
class MoreB<
    TA extends MoreA<TA,TB,TC>, 
    TB extends MoreB<TA,TB,TC>, 
    TC extends MoreC<TA,TB,TC>
> extends B<TA, TB, TC> { ... } 
class MoreC<
    TA extends MoreA<TA,TB,TC>, 
    TB extends MoreB<TA,TB,TC>, 
    TC extends MoreC<TA,TB,TC>
> extends C<TA, TB, TC> { ... } 

//terminal coheritance
class GoalA extends MoreA<GoalA, GoalB, GoalC> { ... }
class GoalB extends MoreB<GoalA, GoalB, GoalC> { ... }
class GoalC extends MoreC<GoalA, GoalB, GoalC> { ... }
```

The C# version is a bit less wordy but still is monstrous. These constructs are difficult to read and difficult to understand.

Let's compare with the hypothetical new way of the expressing the same things:

```Typescript
module Base<A,B,C> {
    class A { ... }
    class B { ... }
    class C { ... }
}

module More<MoreA, MoreB, MoreC>
  extends Base<MoreA, MoreB, MoreC>  {
       class MoreA { ... }
       class MoreB { ... }
       class MoreC { ... }
}

module Goal
  extends More<GoalA, GoalB, GoalC> 
  // in fact we can have more `extends` clauses for each relation
{
       class GoalA { ... }
       class GoalB { ... }
       class GoalC { ... }
       ...
}
```

Here each module, in fact, does two things:
- Defines a relation between classes
- Provides an implementation for each class

One relation can require for other relations to be defined on the target classes (like a Prolog rules) as well as require for the initial implementation (the inheritance).

Ok, maybe it's really too complicated for an average JS/TS programmer.
At least it looks much simpler.

Maybe one day, in another language... -)

vladima commented:
yes, it is possible in another language  :) - Scala can do this with  abstract types and explicitly typed self-references
```scala
abstract class Graph {
  // abstract types that should be implemented in derived types
  type Edge <: EdgeContract;
  type Vertex <: VertexContract;
  
  trait EdgeContract {
    def from: Vertex;
    def to: Vertex
  }
  
  trait VertexContract {
    // specify self type for 'this'
    this: Vertex =>
    
    var incoming: List[Edge] = Nil;
    var outgoing: List[Edge] = Nil;
    
    def connect(to: Vertex): Vertex = {
      val edge = newEdge(this, to); // without self type annotation this line will have an error since static type of 'this' is VertexContract not Vertex
      this.outgoing = edge :: this.outgoing;
      to.incoming = edge :: to.incoming;
      to
    }
  }
  
  // abstract members
  def edges: List[Edge];
  def vertexes: List[Vertex];  
  def newEdge(from: Vertex, to:Vertex): Edge;
}

class Geo extends Graph {
  // implement abstract types Edge and Vertex with concrete types
  type Edge = Road;
  type Vertex = City;
  
  var edgeList: List[Edge] = Nil;
  var vertexList: List[Vertex] = Nil;
  // implement abstract members
  def edges = this.edgeList
  def vertexes = this.vertexList;  
  def newEdge(from: City, to: City) = new Road(from, to);
  
  class City(val name: String) extends VertexContract {
  }
  
  class Road(val from: Vertex, val to: Vertex) extends EdgeContract {
  }  
}

val geo = new Geo;
val city1 = new geo.City("City1")
val city2 = new geo.City("City2");
val road = city1.connect(city2); // type of road is Road
println(road.from.name);         // type of road.from is City

```

DanielRosenwasser commented on src/compiler/checker.ts:
```diff
@@ -38,6 +38,7 @@ namespace ts {
         let Signature = objectAllocator.getSignatureConstructor();
 
         let typeCount = 0;
+        let symbolCount = 0;
```
Was this something we just weren't accounting for?

ahejlsberg replied:
Yes, we were only showing the number of symbols created by the binder and not including the transient symbols created by the checker.

DanielRosenwasser commented:
Can you add a contextual typing test? I might've missed it, but I don't see any.

```TypeScript
interface X {
    a: (p: this) => this;
}

interface Y extends X {
}

var x: Y = {
    a(p) {
        return p;
    }
}

var y = x.a(x);
```

ahejlsberg commented:
With latest commits we now report an error when an emitted type in a declaration file references an inaccessible `this` type, as per our design meeting discussion. Also, added a contextual typing test as suggested.

jbondc commented:
For ```this``` inside object literals, a solution could be:
```
interface Thing {
    getResource(): <ResName>{
        methodA(): this<ResName>; // refers to object literal
        methodB(): this; // by default refers to self reference of container type 'Thing'
    }
}
```

mhegazy commented:
Created https://github.com/Microsoft/TypeScript/issues/5038 to re-enable the fourslash tests. As for the declaration emitter changes, i think we can achieve the error reporting in a less hacky manner. will look into that as part of my declaration emit cleanup.

mhegazy commented:
:+1:

mhegazy commented:
:tada:

vilic commented:
:tada: :+1:

mitjap commented:
should this also support this syntax?

    interface A<T> { }
    interface B<T extends A<T>> { }

mhegazy commented:
@mitjap: see #2225

mjohnsonengr commented:
This seems to cause the following issue.  In the example below in the `getProperty()` method, style is assigned to `this`.  This used to be able to infer the type Style, but the latest code including this pull request does not, and the assignment to style at the end of the while loop throws an error, "Type 'Style' is not assignable to type 'this'.

Is this expected?

	class Style {
		public property: number;
		public basedOn: string;
		
		/** if property is null, get from basedOn */
		public getProperty(): number {
			var style = this; // used to infer type of Style
			while (style !== null) {
				if (style.property)
					return style.property;
				// get style referenced by basedOn
				style = style.basedOn === null ? null : getStyle(style.basedOn);
			}
			return null;
		}
	}

RyanCavanaugh commented:
@mjohnsonengr this is the intended behavior. If there were a subclass of `Style`, the result of `getStyle` would be unassignable to a variable of type `this` (i.e. attempting to assign a `Style` to a `SuperCoolStyle` during a polymorphic invocation of `getProperty`).

I'd recommend writing `var style: Style = this;` - that will solve the issue

mjohnsonengr commented:
Nevermind, saw the referenced issue, https://github.com/Microsoft/TypeScript/issues/5056

mjohnsonengr commented:
Ahh, but thank your for your reply @RyanCavanaugh! :)

aalpgiray commented:
so what should I do to use this instead of state below

```typescript
function NavigatableRecordCreator(defaultValues: { [key: string]: any; }, name?: string) {
    abstract class NavigatableRecord<P extends NavigatableRecord<P>> extends Record(defaultValues, name) {
        SetValue<T>(fn: (x: NavigableObject<P>) => NavigableObject<T>, value: T, state: P) {
            return this.setIn(fn(new NavigableObject<P>(state)).getPath(), value) // can I use this instead of state here
        }
    }
    return NavigatableRecord;
}
````

shelby3 commented:
@ahejlsberg [wrote](https://github.com/Microsoft/TypeScript/pull/4910#issue-107614930):
> Each subclass of the above `Entity` will have a `clone` method that returns an instance of the subclass, and an `equals` method that takes another instance of the subclass.

> The `this` type is a subtype of and assignable to the instance type of the containing class or interface, _but not_ vice-versa (because `this` might actually be a subclass).

As [I wrote](https://github.com/Microsoft/TypeScript/issues/3694#issuecomment-246097444) in #3694, how can `clone(): this` be implemented since it is not legal to assign an instance of the subclass to `this`???
